### PR TITLE
Add const generic descriptors to hipsparse

### PR DESCRIPTION
**Pull request by Angeloo01**
resolves [SWDEV-374035](https://ontrack-internal.amd.com/browse/SWDEV-374035)

Summary of proposed changes:
 - Add const_spvec_descr, const_spmat_descr, const_dnvec_descr, const_dnmat_descr
 - Add and update corresponding auxiliary functions to accommodate new descriptors
 - create pre and post cuda 12 versions of affected auxiliary routines with cuda backend

Have yet to build and test on cuda 12 due to removed functions from cuda not yet propagated to hip. Builds and passes on cuda 11.4

### CODE REVIEW

File: clients/include/testing_const_dnmat_descr.hpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
+/* ************************************************************************
+ * Copyright (C) 2020 Advanced Micro Devices, Inc. All rights Reserved.
```
Comment (jsandham) [inline clients/include/testing_const_dnmat_descr.hpp:2]: 2023

File: clients/include/testing_const_dnvec_descr.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
+/* ************************************************************************
+ * Copyright (C) 2020 Advanced Micro Devices, Inc. All rights Reserved.
```
Comment (jsandham) [inline clients/include/testing_const_dnvec_descr.hpp:2]: 2023

**Review by jsandham — APPROVED** [review]

File: clients/include/testing_const_dnvec_descr.hpp (hunk 2 above)
Comment (Angeloo01) [reply clients/include/testing_const_dnvec_descr.hpp]: fixed

File: clients/include/testing_const_dnmat_descr.hpp (hunk 1 above)
Comment (Angeloo01) [reply clients/include/testing_const_dnmat_descr.hpp]: fixed

**Comment by ntrost57** [comment]
cusparse did not deprecated any of those functions? They just replaced the non const version with const? I mean, this will break existing applications, no?

**Comment by Angeloo01** [comment]
> cusparse did not deprecated any of those functions? They just replaced the non const version with const? I mean, this will break existing applications, no?

Passing in a non-const descriptor into a const descriptor parameter will work. So existing applications with the original descriptors would still work with the new functions. If that's what you mean?

**Comment by ntrost57** [comment]
> > cusparse did not deprecated any of those functions? They just replaced the non const version with const? I mean, this will break existing applications, no?
> 
> Passing in a non-const descriptor into a const descriptor parameter will work. So existing applications with the original descriptors would still work with the new functions. If that's what you mean?

Yes. Guess I missed it's just a typedef.

**Review by ntrost57 — APPROVED** [review]

**Review by YvanMokwinski — APPROVED** [review]

**Pull request merged**